#include "setupPose.h"

extern bool measureStartBtn;
extern bool imported;

setupPose::setupPose(QWidget *parent)
	: QDialog(parent, Qt::FramelessWindowHint), ui(new Ui::setupPose)
{
	ui->setupUi(this);
	this->initCamera();
	this->cameraDeviceSearch();
	
	QComboBox* comboBox = ui->comboBox;
	QPushButton* startBtn = ui->startBtn;
	QPushButton* stopBtn = ui->stopBtn;
	QPushButton* captureBtn = ui->captureBtn;

	comboBox->hide();
	startBtn->hide();
	stopBtn->hide();
	captureBtn->hide();

	//window drag and drop move
	justOneCount = 0;
	mouseX = this->geometry().x();
	mouseY = this->geometry().y();
	absY = this->geometry().y();
	absX = this->geometry().x();

	initBtn = ui->initBtn;
	initBtn->setEnabled(false);

	count = ui->count;

	this->onStartBtn();
}

setupPose::~setupPose()
{
	delete myCamera;
	delete myCapture;
	delete ui;
	
}

void setupPose::initCamera()
{
	myCamera = new QCamera;
	myCapture = new QCameraImageCapture(myCamera);
	myViewfinder = new QCameraViewfinder();

	QVBoxLayout* vLay = new QVBoxLayout();
	vLay->addWidget(myViewfinder);
	ui->camOut->setLayout(vLay);
}

void setupPose::cameraDeviceSearch()
{
	camDevNameLists.clear();
	ui->comboBox->clear();

	for each (const QByteArray &deviceName in QCamera::availableDevices())
	{
		QString description;
		description = myCamera->deviceDescription(deviceName);
		camDevNameLists.append(deviceName);
		ui->comboBox->addItem(description);
	}
}

void setupPose::onStartBtn()
{
	delete myCapture;
	delete myCamera;
	if (camDevNameLists.count() < 1) {
		myCamera = new QCamera;
	}
	else {
		int curIndex = ui->comboBox->currentIndex();
		myCamera = new QCamera(camDevNameLists.at(curIndex));
	}

	connect(myCamera, SIGNAL(error(QCamera::Error)), this, SLOT(camError(QCamera::Error)));

	myCamera->setViewfinder(myViewfinder);
	myCamera->setCaptureMode(QCamera::CaptureVideo);
	myCapture = new QCameraImageCapture(myCamera);

	myCapture->setCaptureDestination(QCameraImageCapture::CaptureToBuffer | QCameraImageCapture::CaptureToFile);
	myCapture->setBufferFormat(QVideoFrame::Format_RGB32);

	connect(myCapture, SIGNAL(imageCaptured(int, QImage)), this, SLOT(imageCaptured(int, QImage)));


	initBtn->setEnabled(true); 
	myCamera->start();
	
}

void setupPose::onStopBtn()
{
	myCamera->stop();
}

void setupPose::onCaptureBtn()
{
	myCapture->capture("c:\CuptureImage.jpg");
}

void setupPose::camError(QCamera::Error error)
{
	qDebug() << Q_FUNC_INFO << "ERROR : " << error;
}

void setupPose::imageCapture(int pid, QImage pPriview)
{
	Q_UNUSED(pid);

	qDebug() << "IMAGE CAPTURE SIZE (WIDTH X HEIGHT) : " << pPriview.byteCount();
}

void setupPose::initPoseBtn()
{
	measureStartBtn = true;
}

void setupPose::mouseMoveEvent(QMouseEvent* mouse)
{
	if (this->isMaximized() == true) //ï¿½Ö´ï¿½È­ ï¿½Ç¾ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿?ï¿½ï¿½ï¿½ï¿½
		return;

	if (mouse->button() == Qt::RightButton) //ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½Å¬ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿?ï¿½ï¿½ï¿½ï¿½
		return;

	mouseX = QCursor::pos().x(); //ï¿½ï¿½ï¿½ì½º ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½Ç¥
	mouseY = QCursor::pos().y();

	if (justOneCount == 0)
	{
		absX = mouse->pos().x(); //ï¿½ï¿½ï¿½ì½º ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½?ï¿½ï¿½ï¿½ï¿½
		absY = mouse->pos().y();
		justOneCount++; //1ï¿½ÌµÇ¸ï¿½ ï¿½ï¿½ ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½ ï¿½ï¿½ï¿½ï¿½
	}
	this->move(mouseX - absX, mouseY - absY); //ï¿½ï¿½ï¿½ï¿½ï¿½ï¿½Ç¥ï¿½ï¿½ï¿½ï¿½ ï¿½ï¿½ï¿½ï¿½ï¿½Ç¥ï¿½ï¿?ï¿½ï¿½ï¿½ï¿½ ï¿½Ìµï¿½ï¿½Ï´ï¿½ ï¿½ï¿½ï¿½ï¿½
}

void setupPose::mouseReleaseEvent(QMouseEvent*)
{
	justOneCount = 0; //ï¿½ï¿½ï¿½ì½ºï¿½ï¿½ Å¬ï¿½ï¿½ ï¿½ï¿½ï¿½ï¿½ï¿½Ï¸ï¿½ ï¿½Ù½ï¿½ 0ï¿½ï¿½ï¿½ï¿½ï¿½Ï¿ï¿½ ï¿½Ýºï¿½ï¿½ï¿½ë°¡ï¿½ï¿?
}

void setupPose::initClose()
{
	myCamera->stop();
	this->hide();
	delete ui;
}

void setupPose::setCount(int count)
{	
	//if complete immediate end

	//else not complete re shoot
}
